ITensorGaussianMPS/src/gmps.jl

```diff
@@ -188,6 +294,193 @@ function givens_rotations(v::AbstractVector{ElT}) where {ElT}
   return gs, r
 end
 
+givens_rotations(v::ConservingNf) = return givens_rotations(v.data)
+
+"""
+  givens_rotations(_v0::Pairing)
+  
+  For a vector
+  ```julia
+  v=_v0.data
+  ```
+  from a fermionic Gaussian state, return the `4*length(v)-1`
+  real Givens/Boguliobov rotations `g` and the norm `r` such that:
+  ```julia
+  g * v ≈ r * [n == 2 ? 1 : 0 for n in 1:length(v)]
+ c
+  with `g` being composed of diagonal rotation aligning pairs
+  of complex numbers in the complex plane, and Givens/Boguliobov Rotations
+  with real arguments only, acting on the interlaced single-particle space of
+  annihilation and creation operator coefficients.
+  """
+function givens_rotations(_v0::Pairing;)
+  v0 = _v0.data
+  N = div(length(v0), 2)
+  if N == 1
+    error(
+      "Givens rotation on 2-element vector not allowed for Pairing-type calculations. This should have been caught elsewhere.",
+    )
+  end
+  ElT = eltype(v0)
+  gs = Circuit{ElT}([])
+  v = copy(v0)
+  r = v[2]
+  ##Given's rotations from creation-operator coefficients
+  gscc, _ = givens_rotations(v[2:2:end])
+  gscc = scale!(gscc, 2)
+  gsca = Circuit(copy(gscc.rotations))
+  gsca = shift!(gsca, -1)
+  gsca = conj!(gsca)
+  gsc = interleave(gscc, gsca)
+  v = gsc * v
+  LinearAlgebra.lmul!(gsc, gs)
+
+  ##Given's rotations from annihilation-operator coefficients
+  gsaa, _ = givens_rotations(v[3:2:end])
+  gsaa = scale!(gsaa, 2)
+  gsaa = shift!(gsaa, +1)
+  gsac = Circuit(copy(gsaa.rotations))
+  gsac = shift!(gsac, +1)
+  gsac = conj!(gsac)
+  gsa = interleave(gsac, gsaa)
+  v = gsa * v
+  LinearAlgebra.lmul!(gsa, gs)
+
+  ##Boguliobov rotation for remaining Bell pair
+  g1, r = givens(v, 2, 3)
+  g2 = Givens(1, 4, g1.c, g1.s')
+  v = g1 * v
+  v = g2 * v #should have no effect
+  LinearAlgebra.lmul!(g2, gs)
+  LinearAlgebra.lmul!(g1, gs)
+  return gs, r
+end
+
+function check_pairing_correlations(Λ0::AbstractMatrix{ElT}) where {ElT<:Number}
+  paired = false
+  Λblocked = reverse_interleave(Λ0)
+  N = div(size(Λblocked, 1), 2)
+  if all(abs.(Λblocked[1:N, (N + 1):end]) .<= eps(Float64))
```

Comment:
It would be faster to do:
```julia
all(x -> abs(x) <= eps(eltype(Λ0)), @view Λblocked[1:N, (N + 1):end])
```
which won't allocated temporary matrices.

Reply:
Ok. What about the return signature here for AbstractMatrices and AbstractSymmetry types? The former returns ::bool,::AbstractMatrix, the latter returns ::AbstractSymmetry --- probably best to go with ::bool,::Any in both cases?

Reply:
What if they both returned `AbstractSymmetry`? It seems like that covers the purpose of the `Bool` anyway.